### PR TITLE
Replace RuntimeException with CakeException, DatabaseException or InvalidArgumentException

### PR DESCRIPTION
**Pull request by othercorey**
The changes in tests are just updating the matching expected exceptions.

All of the changes to InvalidArgumentException have been marked.

### CODE REVIEW

File: src/Collection/CollectionTrait.php [hunk 1]
```diff
@@ -750,7 +749,7 @@ public function listNested(
             ];
 
             if (!isset($modes[$order])) {
-                throw new RuntimeException(sprintf(
+                throw new InvalidArgumentException(sprintf(
```
Comment (othercorey) [inline src/Collection/CollectionTrait.php:752]: This changed to InvalidArgumentException

File: src/Command/Helper/ProgressHelper.php [hunk 2]
```diff
@@ -76,7 +76,7 @@ public function output(array $args): void
             $args['callback'] = $args[0];
         }
         if (!$args['callback'] || !is_callable($args['callback'])) {
-            throw new RuntimeException('Callback option must be a callable.');
+            throw new InvalidArgumentException('Callback option must be a callable.');
```
Comment (othercorey) [inline src/Command/Helper/ProgressHelper.php:79]: This changed to InvalidArgumentException

File: src/Console/CommandRunner.php [hunk 3]
```diff
@@ -126,10 +125,14 @@ public function setAliases(array $aliases)
      * @param array $argv The arguments from the CLI environment.
      * @param \Cake\Console\ConsoleIo|null $io The ConsoleIo instance. Used primarily for testing.
      * @return int The exit code of the command.
-     * @throws \RuntimeException
+     * @throws \InvalidArgumentException
      */
     public function run(array $argv, ?ConsoleIo $io = null): int
     {
+        if (empty($argv)) {
+            throw new InvalidArgumentException('Cannot run any commands. No arguments received.');
```
Comment (othercorey) [inline src/Console/CommandRunner.php:133]: This changed to InvalidArgumentException

File: src/Console/TestSuite/ConsoleIntegrationTestTrait.php [hunk 4]
```diff
@@ -89,7 +88,9 @@ public function exec(string $command, array $input = []): void
         if ($this->_in === null) {
             $this->_in = new StubConsoleInput($input);
         } elseif ($input) {
-            throw new RuntimeException('You can use `$input` only if `$_in` property is null and will be reset.');
+            throw new InvalidArgumentException(
```
Comment (othercorey) [inline src/Console/TestSuite/ConsoleIntegrationTestTrait.php:91]: This changed to InvalidArgumentException

File: src/Database/Driver/TupleComparisonTranslatorTrait.php [hunk 5]
```diff
@@ -58,7 +58,7 @@ protected function _transformTupleComparison(TupleComparison $expression, Query
 
         $operator = strtoupper($expression->getOperator());
         if (!in_array($operator, ['IN', '='])) {
-            throw new RuntimeException(
+            throw new InvalidArgumentException(
```
Comment (othercorey) [inline src/Database/Driver/TupleComparisonTranslatorTrait.php:61]: This changed to InvalidArgumentException

File: src/Database/Expression/OrderByExpression.php [hunk 6]
```diff
@@ -75,7 +75,7 @@ protected function _addConditions(array $conditions, array $types): void
                 is_string($val) &&
                 !in_array(strtoupper($val), ['ASC', 'DESC'], true)
             ) {
-                throw new RuntimeException(
+                throw new InvalidArgumentException(
```
Comment (othercorey) [inline src/Database/Expression/OrderByExpression.php:78]: This changed to InvalidArgumentException

File: src/Database/Query.php [hunk 7]
```diff
@@ -1711,12 +1711,12 @@ public function unionAll(Query|string $query, bool $overwrite = false)
...
      * @param array<string, string> $types A map between columns & their datatypes.
      * @return $this
-     * @throws \RuntimeException When there are 0 columns.
+     * @throws \InvalidArgumentException When there are 0 columns.
      */
     public function insert(array $columns, array $types = [])
     {
         if (empty($columns)) {
-            throw new RuntimeException('At least 1 column is required to perform an insert.');
+            throw new InvalidArgumentException('At least 1 column is required to perform an insert.');
```
Comment (othercorey) [inline src/Database/Query.php:1719]: This changed to InvalidArgumentException

File: src/Error/Debug/ConsoleFormatter.php [hunk 8]
```diff
@@ -138,7 +138,7 @@ protected function export(NodeInterface $var, int $indent): string
         if ($var instanceof SpecialNode) {
             return $this->style('special', $var->getValue());
         }
-        throw new RuntimeException('Unknown node received ' . get_class($var));
+        throw new InvalidArgumentException('Unknown node received ' . get_class($var));
```
Comment (othercorey) [inline src/Error/Debug/ConsoleFormatter.php:141]: This changed to InvalidArgumentException

File: src/Error/Debug/HtmlFormatter.php [hunk 9]
```diff
@@ -143,7 +143,7 @@ protected function export(NodeInterface $var, int $indent): string
         if ($var instanceof SpecialNode) {
             return $this->style('special', $var->getValue());
         }
-        throw new RuntimeException('Unknown node received ' . get_class($var));
+        throw new InvalidArgumentException('Unknown node received ' . get_class($var));
```
Comment (othercorey) [inline src/Error/Debug/HtmlFormatter.php:146]: This changed to InvalidArgumentException

File: src/Error/Debug/TextFormatter.php [hunk 10]
```diff
@@ -87,7 +87,7 @@ protected function export(NodeInterface $var, int $indent): string
         if ($var instanceof SpecialNode) {
             return $var->getValue();
         }
-        throw new RuntimeException('Unknown node received ' . get_class($var));
+        throw new InvalidArgumentException('Unknown node received ' . get_class($var));
```
Comment (othercorey) [inline src/Error/Debug/TextFormatter.php:90]: This changed to InvalidArgumentException

File: src/Error/Debugger.php [hunk 11]
```diff
@@ -201,7 +201,7 @@ public static function setEditor(string $name): void
         $instance = static::getInstance();
         if (!isset($instance->editors[$name])) {
             $known = implode(', ', array_keys($instance->editors));
-            throw new RuntimeException("Unknown editor `{$name}`. Known editors are {$known}");
+            throw new InvalidArgumentException("Unknown editor `{$name}`. Known editors are {$known}");
```
Comment (othercorey) [inline src/Error/Debugger.php:204]: This changed to InvalidArgumentException

File: src/Error/Debugger.php [hunk 12]
```diff
@@ -218,7 +218,7 @@ public static function editorUrl(string $file, int $line): string
         $instance = static::getInstance();
         $editor = $instance->getConfig('editor');
         if (!isset($instance->editors[$editor])) {
-            throw new RuntimeException("Cannot format editor URL `{$editor}` is not a known editor.");
+            throw new InvalidArgumentException("Cannot format editor URL `{$editor}` is not a known editor.");
```
Comment (othercorey) [inline src/Error/Debugger.php:221]: This changed to InvalidArgumentException

File: src/Event/Decorator/ConditionDecorator.php [hunk 13]
```diff
@@ -67,7 +67,7 @@ protected function _evaluateCondition(string $condition, EventInterface $event):
             return $condition !== 'unless';
         }
         if (!is_callable($this->_options[$condition])) {
-            throw new RuntimeException(self::class . ' the `' . $condition . '` condition is not a callable!');
+            throw new InvalidArgumentException(self::class . ' the `' . $condition . '` condition is not a callable!');
```
Comment (othercorey) [inline src/Event/Decorator/ConditionDecorator.php:70]: This changed to InvalidArgumentException

File: src/Http/MiddlewareQueue.php [hunk 14]
```diff
@@ -63,15 +63,15 @@ public function __construct(array $middleware = [])
...
+     * @throws \InvalidArgumentException If Middleware not found.
      */
     protected function resolve(MiddlewareInterface|Closure|string $middleware): MiddlewareInterface
     {
         if (is_string($middleware)) {
             /** @psalm-var class-string<\Psr\Http\Server\MiddlewareInterface>|null $className */
             $className = App::className($middleware, 'Middleware', 'Middleware');
             if ($className === null) {
-                throw new RuntimeException(sprintf(
+                throw new InvalidArgumentException(sprintf(
```
Comment (othercorey) [inline src/Http/MiddlewareQueue.php:74]: This changed to InvalidArgumentException

File: src/ORM/Behavior/TimestampBehavior.php [hunk 15]
```diff
@@ -220,7 +220,7 @@ protected function _updateField(EntityInterface $entity, string $field, bool $re
         $type = TypeFactory::build($columnType);
 
         if (!$type instanceof DateTimeType) {
-            throw new RuntimeException('TimestampBehavior only supports columns of type DateTimeType.');
+            throw new InvalidArgumentException('TimestampBehavior only supports columns of type DateTimeType.');
```
Comment (othercorey) [inline src/ORM/Behavior/TimestampBehavior.php:223]: This changed to InvalidArgumentException

File: src/Routing/RouteBuilder.php [hunk 16]
```diff
@@ -999,7 +998,7 @@ public function applyMiddleware(string ...$names)
             if (!$this->_collection->middlewareExists($name)) {
                 $message = "Cannot apply '$name' middleware or middleware group. " .
                     'Use registerMiddleware() to register middleware.';
-                throw new RuntimeException($message);
+                throw new InvalidArgumentException($message);
```
Comment (othercorey) [inline src/Routing/RouteBuilder.php:1001]: This changed to InvalidArgumentException

File: src/Routing/RouteCollection.php [hunk 17]
```diff
@@ -409,19 +409,19 @@ public function registerMiddleware(string $name, MiddlewareInterface|Closure|str
...
      * @return $this
-     * @throws \RuntimeException
+     * @throws \InvalidArgumentException
      */
     public function middlewareGroup(string $name, array $middlewareNames)
     {
         if ($this->hasMiddleware($name)) {
             $message = "Cannot add middleware group '$name'. A middleware by this name has already been registered.";
-            throw new RuntimeException($message);
+            throw new InvalidArgumentException($message);
```
Comment (othercorey) [inline src/Routing/RouteCollection.php:418]: This changed to InvalidArgumentException

File: src/Routing/RouteCollection.php [hunk 18]
```diff
@@ -409,19 +409,19 @@ public function registerMiddleware(string $name, MiddlewareInterface|Closure|str
...
         }
 
         foreach ($middlewareNames as $middlewareName) {
             if (!$this->hasMiddleware($middlewareName)) {
                 $message = "Cannot add '$middlewareName' middleware to group '$name'. It has not been registered.";
-                throw new RuntimeException($message);
+                throw new InvalidArgumentException($message);
```
Comment (othercorey) [inline src/Routing/RouteCollection.php:424]: This changed to InvalidArgumentException

File: src/Routing/RouteCollection.php [hunk 19]
```diff
@@ -480,7 +480,7 @@ public function getMiddleware(array $names): array
                 continue;
             }
             if (!$this->hasMiddleware($name)) {
-                throw new RuntimeException(sprintf(
+                throw new InvalidArgumentException(sprintf(
```
Comment (othercorey) [inline src/Routing/RouteCollection.php:483]: This changed to InvalidArgumentException

File: src/Utility/CookieCryptTrait.php [hunk 20]
```diff
@@ -83,7 +83,7 @@ protected function _checkCipher(string $encrypt): void
                 'Invalid encryption cipher. Must be one of %s or false.',
                 implode(', ', $this->_validCiphers)
             );
-            throw new RuntimeException($msg);
+            throw new InvalidArgumentException($msg);
```
Comment (othercorey) [inline src/Utility/CookieCryptTrait.php:86]: This changed to InvalidArgumentException

File: src/Utility/Hash.php [hunk 21]
```diff
@@ -486,7 +485,7 @@ public static function combine(
         }
 
         if (is_array($keys) && count($keys) !== count($vals)) {
-            throw new RuntimeException(
+            throw new InvalidArgumentException(
```
Comment (othercorey) [inline src/Utility/Hash.php:488]: This changed to InvalidArgumentException

File: src/Utility/Security.php [hunk 22]
```diff
@@ -69,7 +69,7 @@ public static function hash(string $string, ?string $algorithm = null, string|bo
 
         $availableAlgorithms = hash_algos();
         if (!in_array($algorithm, $availableAlgorithms, true)) {
-            throw new RuntimeException(sprintf(
+            throw new InvalidArgumentException(sprintf(
```
Comment (othercorey) [inline src/Utility/Security.php:72]: This changed to InvalidArgumentException

File: src/Validation/Validation.php [hunk 23]
```diff
@@ -1454,7 +1453,7 @@ public static function geoCoordinate(mixed $value, array $options = []): bool
             'type' => 'latLong',
         ];
         if ($options['type'] !== 'latLong') {
-            throw new RuntimeException(sprintf(
+            throw new InvalidArgumentException(sprintf(
```
Comment (othercorey) [inline src/Validation/Validation.php:1456]: This changed to InvalidArgumentException

File: src/Validation/ValidatorAwareTrait.php [hunk 24]
```diff
@@ -107,14 +107,14 @@ public function getValidator(?string $name = null): Validator
...
-     * @throws \RuntimeException
+     * @throws \InvalidArgumentException
      */
     protected function createValidator(string $name): Validator
     {
         $method = 'validation' . ucfirst($name);
         if (!$this->validationMethodExists($method)) {
             $message = sprintf('The %s::%s() validation method does not exists.', static::class, $method);
-            throw new RuntimeException($message);
+            throw new InvalidArgumentException($message);
```
Comment (othercorey) [inline src/Validation/ValidatorAwareTrait.php:117]: This changed to InvalidArgumentException

File: src/Validation/ValidatorAwareTrait.php [hunk 25]
```diff
@@ -127,7 +127,7 @@ protected function createValidator(string $name): Validator
         }
 
         if (!$validator instanceof Validator) {
-            throw new RuntimeException(sprintf(
+            throw new InvalidArgumentException(sprintf(
```
Comment (othercorey) [inline src/Validation/ValidatorAwareTrait.php:130]: This changed to InvalidArgumentException

File: src/View/Form/EntityContext.php [hunk 26]
```diff
@@ -595,7 +596,7 @@ protected function _getValidator(array $parts): Validator
 
         $table = $this->_getTable($parts);
         if (!$table) {
-            throw new RuntimeException('Validator not found: ' . $key);
+            throw new InvalidArgumentException('Validator not found: ' . $key);
```
Comment (othercorey) [inline src/View/Form/EntityContext.php:599]: This changed to InvalidArgumentException

File: src/View/Helper/FormHelper.php [hunk 27]
```diff
@@ -1209,7 +1208,7 @@ protected function _getInput(string $fieldName, array $options): array|string
 
                 return $this->{$options['type']}($fieldName, $opts, $options + ['label' => $label]);
             case 'input':
-                throw new RuntimeException("Invalid type 'input' used for field '$fieldName'");
+                throw new InvalidArgumentException("Invalid type 'input' used for field '$fieldName'");
```
Comment (othercorey) [inline src/View/Helper/FormHelper.php:1211]: This changed to InvalidArgumentException

File: src/View/StringTemplate.php [hunk 28]
```diff
@@ -230,12 +230,12 @@ public function remove(string $name): void
...
      * @param array<string, mixed> $data The data to insert.
      * @return string Formatted string
-     * @throws \RuntimeException If template not found.
+     * @throws \InvalidArgumentException If template not found.
      */
     public function format(string $name, array $data): string
     {
         if (!isset($this->_compiled[$name])) {
-            throw new RuntimeException("Cannot find template named '$name'.");
+            throw new InvalidArgumentException("Cannot find template named '$name'.");
```
Comment (othercorey) [inline src/View/StringTemplate.php:238]: This changed to InvalidArgumentException

File: src/View/View.php [hunk 29]
```diff
@@ -704,13 +704,13 @@ public function element(string $name, array $data = [], array $options = []): st
...
      * @return string The rendered content.
-     * @throws \RuntimeException When $options is lacking a 'key' option.
+     * @throws \InvalidArgumentException When $options is lacking a 'key' option.
      */
     public function cache(callable $block, array $options = []): string
     {
         $options += ['key' => '', 'config' => $this->elementCache];
         if (empty($options['key'])) {
-            throw new RuntimeException('Cannot cache content with an empty key');
+            throw new InvalidArgumentException('Cannot cache content with an empty key');
```
Comment (othercorey) [inline src/View/View.php:713]: This changed to InvalidArgumentException

File: src/View/ViewBuilder.php [hunk 30]
```diff
@@ -648,7 +648,7 @@ protected function _checkViewVars(mixed &$item, string $key): void
             $item instanceof Closure ||
             $item instanceof PDO
         ) {
-            throw new RuntimeException(sprintf(
+            throw new InvalidArgumentException(sprintf(
```
Comment (othercorey) [inline src/View/ViewBuilder.php:651]: This changed to InvalidArgumentException

File: src/View/Widget/WidgetLocator.php [hunk 31]
```diff
@@ -124,7 +124,7 @@ public function add(array $widgets): void
             }
 
             if (is_object($widget) && !($widget instanceof WidgetInterface)) {
-                throw new RuntimeException(sprintf(
+                throw new InvalidArgumentException(sprintf(
```
Comment (othercorey) [inline src/View/Widget/WidgetLocator.php:127]: This changed to InvalidArgumentException

File: src/View/Widget/WidgetLocator.php [hunk 32]
```diff
@@ -149,13 +149,13 @@ public function add(array $widgets): void
...
      * @return \Cake\View\Widget\WidgetInterface WidgetInterface instance.
-     * @throws \RuntimeException when widget is undefined.
+     * @throws \InvalidArgumentException when widget is undefined.
      */
     public function get(string $name): WidgetInterface
     {
         if (!isset($this->_widgets[$name])) {
             if (empty($this->_widgets['_default'])) {
-                throw new RuntimeException(sprintf('Unknown widget `%s`', $name));
+                throw new InvalidArgumentException(sprintf('Unknown widget `%s`', $name));
```
Comment (othercorey) [inline src/View/Widget/WidgetLocator.php:158]: This changed to InvalidArgumentException

File: src/View/Widget/WidgetLocator.php [hunk 33]
```diff
@@ -194,7 +194,7 @@ protected function _resolveWidget(array|string $config): WidgetInterface
         $class = array_shift($config);
         $className = App::className($class, 'View/Widget', 'Widget');
         if ($className === null) {
-            throw new RuntimeException(sprintf('Unable to locate widget class "%s"', $class));
+            throw new InvalidArgumentException(sprintf('Unable to locate widget class "%s"', $class));
```
Comment (othercorey) [inline src/View/Widget/WidgetLocator.php:197]: This changed to InvalidArgumentException

**Comment by ADmad** [comment]
LGTM. Later on places where `InvalidArgmentException` is thrown for type / class type checks could be replaced with `asset()` calls.

**Pull request merged**